Return descriptive error if user is running Aggregator in v1.x and attempting to upgrade to 2.0

thomasvn opened the pull request:
## What does this PR change?

Return descriptive error if user is running Aggregator in 1.x and attempting to upgrade to 2.0

## Does this PR rely on any other PRs?

It needs an accompanying Docs PR.

## How does this PR impact users? (This is the kind of thing that goes in release notes!)

Users who deployed Aggregator in v1.107 or v1.108 will receive the following Helm error when upgrading to v2.0.

```
Error: UPGRADE FAILED: cannot patch "max-od-aggregator" with kind StatefulSet: StatefulSet.apps "max-od-aggregator" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden
```

This is an expected error due to changes in the Aggregator Helm template in 2.0. However, this error is difficult to understand. This PR attempts to get ahead of this error, and instead return a helpful message to users. Also directing them to additional documentation on how to resolve the issue.

## Links to Issues or tickets this PR addresses or fixes

## What risks are associated with merging this PR? What is required to fully test this PR?

If users are running Aggregator in 1.x but have removed the label `helm.sh/chart` from the StatefulSet, this logic will not catch their scenario. I would guess this is a minority of cases? For those who deploy via GitOps + `helm template`, this label should still exist in their StatefulSet.

## How was this PR tested?

```sh
# Errors when templating. The namespace it's pointed at currently has a v1.108.1 aggregator running.
$ helm template thomasn-agg-upgrade ./cost-analyzer --dry-run=server

Error: execution error at (cost-analyzer/templates/NOTES.txt:4:3): Detected an existing Aggregator StatefulSet in your namespace. Before upgrading to Kubecost 2.0, please `kubectl delete` this Statefulset. Refer to the following documentation for more information: https://docs.kubecost.com/install-and-configure/install/kubecostv2
```

```sh
# Correctly see a failure here with an error message
$ helm upgrade -i thomasn-agg-upgrade ~/kubecost/cost-analyzer-helm-chart/cost-analyzer \
  --version 2.0.0-rc.1 \
  -f values.yaml

Error: UPGRADE FAILED: execution error at (cost-analyzer/templates/NOTES.txt:4:3): Detected an existing Aggregator StatefulSet in your namespace. Before upgrading to Kubecost 2.0, please `kubectl delete` this Statefulset. Refer to the following documentation for more information: https://docs.kubecost.com/install-and-configure/install/kubecostv2

# Manual intervention
$ kubectl delete sts/thomasn-agg-upgrade-aggregator

# Success
$ helm upgrade -i thomasn-agg-upgrade ~/kubecost/cost-analyzer-helm-chart/cost-analyzer \
  --version 2.0.0-rc.1 \
  -f values.yaml

Release "thomasn-agg-upgrade" has been upgraded. Happy Helming!
NAME: thomasn-agg-upgrade
LAST DEPLOYED: Thu Jan 25 13:36:13 2024
NAMESPACE: thomasn-agg-upgrade
STATUS: deployed
REVISION: 13
NOTES:
--------------------------------------------------

Kubecost 2.0.0 has been successfully installed.

Please allow 5-10 minutes for Kubecost to gather metrics.

When configured, cost reconciliation with cloud provider billing data will have a 48 hour delay.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace thomasn-agg-upgrade deployment/thomasn-agg-upgrade-cost-analyzer 9090

Then, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install

# Run the test again
$ helm rollback thomasn-agg-upgrade 1
```

## Have you made an update to documentation? If so, please provide the corresponding PR.

TODO. Needs an accompanying PR update.

thomasvn committed:
Check if user is running Aggregator in v1.x. If so, generate a warning
that this StatefulSet will require manual deletion.

thomasvn commented on cost-analyzer/templates/_helpers.tpl:
```diff
@@ -25,13 +25,32 @@ Set important variables before starting main templates
-    {{ fail "In Kubecost 2.0, there is no such thing as a federated primary. If you are a Federated ETL user, this setting has been removed. Make sure you have kubecostAggregator.deployMethod set to 'statefulset' and federatedETL.federatedCluster set to 'true'." }}
+{{ define "kubecostV2-preconditions" }}
+  {{/*Get all StatefulSets in this namespace. Iterate through them and fail if any include the name "aggregator"*/}}
+  {{ if and (semverCompare "<2.0.0-0" .Chart.Version)}}
```
I need to double check that this `if` statement does what I expect. I want to check if the _current_ Chart version is pre-2.0. Not the version of the chart we're upgrading to.

thomasvn replied:
Unfortunately, `.Chart.Version` returns you the version you are currently deploying. So the logic I have here won't work.

michaelmdresser replied:
I think Helm attaches metadata about the current release to some object in the cluster, but I'm not sure in what form. This command has to be accessing _something_ in the cluster that should give you the current version in-cluster:
```
→ helm get metadata --namespace kubecost-cloudcost kubecost-cloudcost
NAME: kubecost-cloudcost
CHART: cost-analyzer
VERSION: 2.0.0
APP_VERSION: development
NAMESPACE: kubecost-cloudcost
REVISION: 42
STATUS: deployed
DEPLOYED_AT: 2024-01-24T10:40:35-08:00
```

thomasvn committed:
Remove duplicate check

michaelmdresser reviewed (commented):
What happens with this test:
- Install Helm chart 2.0.0-rc.X with `--set kubecostAggregator.deployMethod=statefulset`
- Upgrade Helm chart with 2.0.0-rc.x with `--set kubecostAggregator.deployMethod=statefulset`
- Basically, upgrade within the same version -- should succeed

chipzoller commented:
Helm creates a Secret which contains metadata and values about the current release. We only need to check for the destination version at this time because it's the second major release and will only be present in the 2.0 branch.

michaelmdresser commented:
@thomasvn My question was to make sure we don't block users going from 2.0 -> 2.0 accidentally with the logic, its less important right now.

It appears that we are labeling cost-analyzer with the Helm chart version:
```
→ k get deployment -n kubecost-cloudcost kubecost-cloudcost-cost-analyzer -o json | jq '.metadata.labels'
{
  "app": "cost-analyzer",
  "app.kubernetes.io/instance": "kubecost-cloudcost",
  "app.kubernetes.io/managed-by": "Helm",
  "app.kubernetes.io/name": "cost-analyzer",
  "helm.sh/chart": "cost-analyzer-2.0.0"
}
```

Can that label be retrieved with `lookup` and then parsed? I realize it might be hard/impossible, but Helm does seem to provide a lot of capabilities with these template functions.

thomasvn commented:
Good catch! Yes this label can be retrieved with `lookup`. Working on this now.

thomasvn committed:
Pivot. Instead of checking .Chart.Version, check the StatefulSet's labels

thomasvn commented:
@michaelmdresser I've updated the logic here to look for a StatefulSet's labels. I've also updated the PR description with the tests I ran. Let me know what you think!

jessegoodier commented:
Nice work, tested, works as expected.

thomasvn committed:
Add `-` to every line of the kubecostV2-preconditions, so that there are not a bunch of spaces in the NOTES.txt file

JesseGoodier reviewed (approved):
works
`
Error: UPGRADE FAILED: execution error at (cost-analyzer/templates/NOTES.txt:4:4): Detected an existing Aggregator StatefulSet in your namespace. Before upgrading to Kubecost 2.0, please `kubectl delete` this Statefulset. Refer to the following documentation for more information: https://docs.kubecost.com/install-and-configure/install/kubecostv2
`

michaelmdresser reviewed (approved):
Nicely done

thomasvn commented:
Merging now! Will follow up with a docs PR later today.

michaelmdresser commented:
/cherry-pick v2.0
